extensions/aws/processors/PutS3Object.cpp

```diff
@@ -0,0 +1,495 @@
+    logger_->log_error("Proxy port invalid");
+    return false;
+  }
+  context->getProperty(ProxyUsername, proxy.username, flow_file);
+  context->getProperty(ProxyPassword, proxy.password, flow_file);
+  if (!proxy.host.empty()) {
+    s3_wrapper_->setProxy(proxy);
+    logger_->log_info("Proxy for PutS3Object was set.");
+  }
+  return true;
+}
+
+void PutS3Object::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) {
+  if (!context->getProperty(Bucket.getName(), put_s3_request_params_.bucket) || put_s3_request_params_.bucket.empty()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Bucket property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Bucket [%s]", put_s3_request_params_.bucket);
+
+  if (!context->getProperty(StorageClass.getName(), put_s3_request_params_.storage_class)
+      || put_s3_request_params_.storage_class.empty()
+      || STORAGE_CLASSES.find(put_s3_request_params_.storage_class) == STORAGE_CLASSES.end()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Storage Class property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Storage Class [%s]", put_s3_request_params_.storage_class);
+
+  std::string value;
+  if (!context->getProperty(Region.getName(), value) || value.empty() || REGIONS.find(value) == REGIONS.end()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Region property missing or invalid");
+  }
+  s3_wrapper_->setRegion(value);
+  logger_->log_debug("PutS3Object: Region [%s]", value);
+
+  uint64_t timeout_val;
+  if (context->getProperty(CommunicationsTimeout.getName(), value) && !value.empty() && core::Property::getTimeMSFromString(value, timeout_val)) {
+    s3_wrapper_->setTimeout(timeout_val);
+    logger_->log_debug("PutS3Object: Communications Timeout [%d]", timeout_val);
+  } else {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Communications Timeout missing or invalid");
+  }
+
+  if (!context->getProperty(ServerSideEncryption.getName(), put_s3_request_params_.server_side_encryption)
+      || put_s3_request_params_.server_side_encryption.empty()
+      || SERVER_SIDE_ENCRYPTIONS.find(put_s3_request_params_.server_side_encryption) == SERVER_SIDE_ENCRYPTIONS.end()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Server Side Encryption property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Server Side Encryption [%s]", put_s3_request_params_.server_side_encryption);
+
+  fillUserMetadata(context);
+}
+
+std::string PutS3Object::parseAccessControlList(const std::string &comma_separated_list) const {
+  std::string result_list;
+  bool is_first = true;
+  for (const auto& user : minifi::utils::StringUtils::split(comma_separated_list, ",")) {
+    if (is_first) {
+      is_first = false;
+    } else {
+      result_list += ", ";
```

Comment:
This is already implemented in join. 

I bit more simple way of doing this:
```
auto users : minifi::utils::StringUtils::split(comma_separated_list, ",")
for(auto& user: users) {
      auto trimmed_user = minifi::utils::StringUtils::trim(user);
    if (trimmed_user.find('@') != std::string::npos) {
      user = "emailAddress=\"" + trimmed_user + "\"";
    } else {
      user += "id=" + trimmed_user;
    }
}
return minifi::utils::StringUtils::join(", ", users);
```

Reply:
Fixed in [e994558](https://github.com/apache/nifi-minifi-cpp/pull/915/commits/e994558ba8904488b0e3489d51236fb7d4e31ce1)